File: sesa-backend/src/main/java/com/sesa/salud/service/VideoconsultaService.java
/**
 * Servicio de videoconsulta — salas y señalización en memoria (WebRTC).
 * Autor: Ing. J Sebastian Vargas S
 */
package com.sesa.salud.service;

import com.sesa.salud.dto.CrearSalaVideoconsultaRequestDto;
import com.sesa.salud.dto.SalaVideoconsultaDto;
import com.sesa.salud.dto.SignalingEventDto;
import com.sesa.salud.dto.SignalingEventWithIndexDto;
import com.sesa.salud.exception.VideoconsultaTokenInvalidoException;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

@Service
public class VideoconsultaService {

    private final Map<String, SalaInfo> salas = new ConcurrentHashMap<>();
    private final Map<String, List<SignalingEventWithIndexDto>> signalingBySala = new ConcurrentHashMap<>();
    private final Map<String, AtomicInteger> signalingIndexBySala = new ConcurrentHashMap<>();

    public SalaVideoconsultaDto crearSala(CrearSalaVideoconsultaRequestDto request) {
        if (request == null) {
            request = new CrearSalaVideoconsultaRequestDto();
        }
        Long citaId = request.getCitaId();
        Long profesionalId = request.getProfesionalId();
        String salaId = UUID.randomUUID().toString().replace("-", "").substring(0, 12);
        String token = UUID.randomUUID().toString().replace("-", "");
        SalaInfo info = new SalaInfo(salaId, token, citaId, profesionalId, null, false, null, new StringBuilder());
        salas.put(salaId, info);
        signalingBySala.put(salaId, new CopyOnWriteArrayList<>());
        signalingIndexBySala.put(salaId, new AtomicInteger(0));
        return SalaVideoconsultaDto.builder()
                .salaId(salaId)
                .role("creador")
                .token(token)
                .citaId(citaId)
                .profesionalId(profesionalId)
                .build();
    }

    public SalaVideoconsultaDto unirseSala(String salaId, String token) {
        SalaInfo info = salas.get(salaId);
        if (info == null) {
            throw new RuntimeException("Sala no encontrada: " + salaId);
        }
        if (token == null || !token.equals(info.token())) {
            throw new VideoconsultaTokenInvalidoException("Token inválido o expirado");
        }
        return SalaVideoconsultaDto.builder()
                .salaId(salaId)
                .role("participante")
                .token(info.token())
                .citaId(info.citaId())
                .profesionalId(info.profesionalId())
                .pacienteId(info.pacienteId())
                .build();
    }

    public void enviarSignaling(String salaId, String token, SignalingEventDto event) {
        validarTokenSala(salaId, token);
        List<SignalingEventWithIndexDto> list = signalingBySala.get(salaId);
        AtomicInteger idx = signalingIndexBySala.get(salaId);
        int next = idx.incrementAndGet();
        list.add(new SignalingEventWithIndexDto(next, event.getType(), event.getPayload()));
    }

    public List<SignalingEventWithIndexDto> obtenerSignaling(String salaId, String token, int afterIndex) {
        validarTokenSala(salaId, token);
        List<SignalingEventWithIndexDto> list = signalingBySala.get(salaId);
        if (list == null) {
            return List.of();
        }
        return list.stream()
                .filter(e -> e.getIndex() > afterIndex)
                .collect(Collectors.toList());
    }

    private void validarTokenSala(String salaId, String token) {
        SalaInfo info = salas.get(salaId);
        if (info == null) {
            throw new RuntimeException("Sala no encontrada: " + salaId);
        }
        if (token == null || !token.equals(info.token())) {
            throw new VideoconsultaTokenInvalidoException("Token inválido o expirado");
        }
    }

    /** Creador: genera token para el asistente (solo toma de notas). Requiere token de sala. */
    public String habilitarAsistente(String salaId, String token) {
        SalaInfo info = salas.get(salaId);
        if (info == null) throw new RuntimeException("Sala no encontrada: " + salaId);
        if (!token.equals(info.token())) throw new VideoconsultaTokenInvalidoException("Token inválido");
        if (info.tokenAsistente() != null) return info.tokenAsistente();
        String tokenAsistente = UUID.randomUUID().toString().replace("-", "");
        SalaInfo updated = new SalaInfo(info.salaId(), info.token(), info.citaId(), info.profesionalId(), info.pacienteId(), info.consentimientoAsistente(), tokenAsistente, info.notas());
        salas.put(salaId, updated);
        return tokenAsistente;
    }

    /** Paciente: registra consentimiento para que un asistente tome notas. Requiere token de sala. */
    public void registrarConsentimientoAsistente(String salaId, String token) {
        SalaInfo info = salas.get(salaId);
        if (info == null) throw new RuntimeException("Sala no encontrada: " + salaId);
        if (!token.equals(info.token())) throw new VideoconsultaTokenInvalidoException("Token inválido");
        SalaInfo updated = new SalaInfo(info.salaId(), info.token(), info.citaId(), info.profesionalId(), info.pacienteId(), true, info.tokenAsistente(), info.notas());
        salas.put(salaId, updated);
    }

    /** Asistente: guarda el contenido completo de las notas. Requiere token de asistente y consentimiento previo. */
    public void guardarNotas(String salaId, String token, String texto) {
        SalaInfo info = salas.get(salaId);
        if (info == null) throw new RuntimeException("Sala no encontrada: " + salaId);
        if (!Boolean.TRUE.equals(info.consentimientoAsistente()) || !token.equals(info.tokenAsistente())) {
            throw new VideoconsultaTokenInvalidoException("Sin consentimiento o token de asistente inválido");
        }
        StringBuilder notas = new StringBuilder(texto != null ? texto : "");
        SalaInfo updated = new SalaInfo(info.salaId(), info.token(), info.citaId(), info.profesionalId(), info.pacienteId(), true, info.tokenAsistente(), notas);
        salas.put(salaId, updated);
    }

    /** Creador: obtiene el resumen (notas del asistente) de la reunión. */
    public String obtenerNotas(String salaId, String token) {
        SalaInfo info = salas.get(salaId);
        if (info == null) throw new RuntimeException("Sala no encontrada: " + salaId);
        if (!token.equals(info.token())) throw new VideoconsultaTokenInvalidoException("Token inválido");
        return info.notas().toString();
    }

    /** Valida si el token es de asistente y hay consentimiento (para permitir acceso a la vista de notas). */
    public boolean validarAsistente(String salaId, String token) {
        SalaInfo info = salas.get(salaId);
        if (info == null) return false;
        return Boolean.TRUE.equals(info.consentimientoAsistente()) && token != null && token.equals(info.tokenAsistente());
    }

    /** Paciente: indica si el profesional solicitó un asistente y aún no ha dado consentimiento. */
    public boolean solicitudAsistentePendiente(String salaId, String token) {
        SalaInfo info = salas.get(salaId);
        if (info == null) return false;
        if (!token.equals(info.token())) return false;
        return info.tokenAsistente() != null && !info.consentimientoAsistente();
    }

    private record SalaInfo(String salaId, String token, Long citaId, Long profesionalId, Long pacienteId,
                            boolean consentimientoAsistente, String tokenAsistente, StringBuilder notas) {}
}

File: sesa-salud/src/app/core/services/videoconsulta.service.ts
/**
 * Servicio de Telemedicina — Videoconsulta WebRTC.
 * Gestiona salas, señalización (REST polling) y conexión peer-to-peer.
 * Autor: Ing. J Sebastian Vargas S
 */

import { Injectable, inject, signal, computed } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, interval, Subject, firstValueFrom, throwError } from 'rxjs';
import { switchMap, takeWhile, tap } from 'rxjs/operators';
import { environment } from '../../../environments/environment';

export interface SalaVideoconsultaDto {
  salaId: string;
  role: 'creador' | 'participante';
  token: string;
  citaId?: number;
  profesionalId?: number;
  pacienteId?: number;
}

export interface SignalingEventDto {
  type: 'offer' | 'answer' | 'ice';
  payload: string; // SDP o ICE candidate JSON
}

@Injectable({ providedIn: 'root' })
export class VideoconsultaService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.apiUrl}/videoconsulta`;

  private readonly _estado = signal<'idle' | 'creando' | 'uniendo' | 'conectando' | 'conectado' | 'error'>('idle');
  private readonly _error = signal<string | null>(null);
  private readonly _salaId = signal<string | null>(null);
  private readonly _role = signal<'creador' | 'participante' | null>(null);
  /** Token de invitación: se guarda al crear o unirse; se envía en unirse y en todas las llamadas de signaling. */
  private readonly _token = signal<string | null>(null);

  readonly estado = this._estado.asReadonly();
  readonly error = this._error.asReadonly();
  readonly salaId = this._salaId.asReadonly();
  readonly role = this._role.asReadonly();
  readonly token = this._token.asReadonly();
  readonly isConectado = computed(() => this._estado() === 'conectado');

  /** Crea una sala de videoconsulta (médico). Opcional: citaId para vincular. */
  crearSala(citaId?: number, profesionalId?: number): Observable<SalaVideoconsultaDto> {
    this._estado.set('creando');
    this._error.set(null);
    return this.http.post<SalaVideoconsultaDto>(`${this.apiUrl}/salas`, { citaId, profesionalId }).pipe(
      tap((s) => {
        this._salaId.set(s.salaId);
        this._role.set(s.role);
        this._token.set(s.token ?? null);
        this._estado.set('idle');
      })
    );
  }

  /** Unirse a una sala existente con el token del enlace (paciente o invitado). */
  unirseSala(salaId: string, token: string): Observable<SalaVideoconsultaDto> {
    this._estado.set('uniendo');
    this._error.set(null);
    this._salaId.set(salaId);
    return this.http.get<SalaVideoconsultaDto>(`${this.apiUrl}/salas/${salaId}`, {
      params: { token },
    }).pipe(
      tap((s) => {
        this._role.set(s.role ?? 'participante');
        this._token.set(s.token ?? token);
        this._estado.set('idle');
      })
    );
  }

  /** Envía un evento de señalización (offer, answer o ice). Token: el del servicio o el pasado explícitamente. */
  enviarSignaling(salaId: string, event: SignalingEventDto, token?: string | null): Observable<void> {
    const t = token ?? this._token();
    if (!t) {
      return throwError(() => new Error('Falta token de sala'));
    }
    if (!this._token()) {
      this._token.set(t);
    }
    return this.http.post<void>(`${this.apiUrl}/salas/${salaId}/signaling`, event, {
      params: { token: t },
    });
  }

  /** Obtiene eventos de señalización desde un índice (polling). Token: el del servicio o el pasado explícitamente. */
  obtenerSignaling(salaId: string, afterIndex: number, token?: string | null): Observable<{ events: Array<SignalingEventDto & { index: number }> }> {
    const t = token ?? this._token();
    if (!t) {
      return throwError(() => new Error('Falta token de sala'));
    }
    if (!this._token()) {
      this._token.set(t);
    }
    return this.http.get<{ events: Array<SignalingEventDto & { index: number }> }>(
      `${this.apiUrl}/salas/${salaId}/signaling`,
      { params: { after: afterIndex.toString(), token: t } }
    );
  }

  /** Limpia estado local (al colgar). */
  limpiarSala(): void {
    this._estado.set('idle');
    this._error.set(null);
    this._salaId.set(null);
    this._role.set(null);
    this._token.set(null);
  }

  /** Construye la URL para compartir (incluye room y token). */
  getUrlUnirse(salaId: string, token: string): string {
    const base = typeof window !== 'undefined' ? window.location.origin : '';
    return `${base}/videoconsulta?room=${encodeURIComponent(salaId)}&token=${encodeURIComponent(token)}`;
  }

  /** Creador: habilita asistente y devuelve token para el enlace. */
  habilitarAsistente(salaId: string, token: string): Observable<{ tokenAsistente: string }> {
    return this.http.post<{ tokenAsistente: string }>(`${this.apiUrl}/salas/${salaId}/habilitar-asistente`, null, {
      params: { token },
    });
  }

  /** Paciente: registra consentimiento para que un asistente tome notas. */
  registrarConsentimientoAsistente(salaId: string, token: string): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/salas/${salaId}/consentimiento-asistente`, null, {
      params: { token },
    });
  }

  /** Asistente: guarda las notas (contenido completo). */
  guardarNotas(salaId: string, token: string, texto: string): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/salas/${salaId}/notas`, { texto }, {
      params: { token },
    });
  }

  /** Creador: obtiene el resumen (notas) de la reunión. */
  obtenerNotas(salaId: string, token: string): Observable<{ notas: string }> {
    return this.http.get<{ notas: string }>(`${this.apiUrl}/salas/${salaId}/notas`, {
      params: { token },
    });
  }

  /** Valida si el token es de asistente y hay consentimiento. */
  validarAsistente(salaId: string, token: string): Observable<{ valido: boolean }> {
    return this.http.get<{ valido: boolean }>(`${this.apiUrl}/salas/${salaId}/validar-asistente`, {
      params: { token },
    });
  }

  /** Paciente: true si el profesional solicitó asistente y aún no ha autorizado. */
  solicitudAsistentePendiente(salaId: string, token: string): Observable<{ pendiente: boolean }> {
    return this.http.get<{ pendiente: boolean }>(`${this.apiUrl}/salas/${salaId}/solicitud-asistente-pendiente`, {
      params: { token },
    });
  }

  /** URL para el asistente (solo toma de notas). */
  getUrlAsistente(salaId: string, tokenAsistente: string): string {
    const base = typeof window !== 'undefined' ? window.location.origin : '';
    return `${base}/videoconsulta/asistente?room=${encodeURIComponent(salaId)}&token=${encodeURIComponent(tokenAsistente)}`;
  }
}
